docs/site/Setting-debug-strings.md

```diff
@@ -0,0 +1,318 @@
+---
+title: "Setting debug strings"
+lang: en
+keywords: LoopBack 4.0, LoopBack 4, base-connector.js
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Setting-debug-strings.html
+summary:
+---
+
+You can specify debug strings when you run an application, as explained below, to display specific log output to the console.
+You can also redirect the output to a file, if desired. These techniques are often helpful in debugging applications.
+
+## Using debug strings
+
+LoopBack has a number of built-in debug strings to help with debugging.Specify a string on the command-line via an environment variable as follows:
+
+**MacOS and Linux**
+
+```shell
+$ DEBUG=<pattern>[,<pattern>...] node .
+```
+
+**Windows**
+
+```shell
+C:\> set DEBUG=<pattern>[,<pattern>...]
+C:\> node .
+```
+
+where &lt;_pattern_&gt; is a string-matching pattern specifying debug strings to match. You can specify as many matching patterns as you wish.
+
+For example (MacOS and Linux):
+
+```shell
+$ DEBUG=loopback:datasource node .
+```
+
+Or, on Windows:
+
+```shell
+C:\> set DEBUG=loopback:datasource
+C:\> node .
+```
+
+You'll see output such as (truncated for brevity):
+
+```
+loopback:datasource Settings: {"name":"db","debug":true} -0ms
+loopback:datasource Settings: {"name":"geo","connector":"rest",...
+```
+
+You can use an asterisk  (`*`) in the pattern to match any string. For example the following would match any debug string containing "oracle":
+
+```shell
+$ DEBUG=*oracle node .
+```
+
+You can also exclude specific debuggers by prefixing them with a "-" character.
+For example, `DEBUG=*,-rest-crud:*` would include all debuggers except those starting with "rest-crud:".
+
+## Debug string format
+
+These strings have the format
+
+`module[:area]:string`
+
+Where:
+
+* _module_ is the name of the module, for example `loopback` or `loopback-connector-mongodb`.
+* _area_ is an optional identifier such as `cli` or `repository` to identify the purpose of the module
+* _string_ is the debug string specified in the target TypeScript/JavaScript source file, such as `application.ts`.
+
+For example:
+
+`loopback:cli:model-generator`
+
+The debug string `model-generator` is specified in the source file [`generators/model/index.js`](https://github.com/strongloop/loopback-next/blob/master/packages/cli/generators/model/index.js) of the `@loopback/cli` module.
+
+## Debug strings reference
+
+<table>
```

Comment:
Did you create this table by hand? Can we automate this?

Reply:
I don't know if we have any script to do it, so I checked modules one by one :flushed: And I didn't include all of the debug strings in here, or the list would be extremely long... I only include those important (from my understanding) ones. Any suggestions?

Reply:
A shell script will do. Or monkey patch the debug function :-)

Reply:
@agnes512 ,

Here's a script I just made:

```sh
#!/bin/bash 
clear
FILE_EXTENSION=$1
SEARCH_TEXT=$2
echo ""
echo " Files with extension : '${FILE_EXTENSION}'  "
echo " that contains text   : '${SEARCH_TEXT}' : "
echo "----------------------------------------------"
echo ""
find . -name "${FILE_EXTENSION}" -exec grep -h "${SEARCH_TEXT}" {} +
echo ""
echo ""




# include file name and found search text
# find . -name "*.ts" -exec grep "debugFactory('loopback" {} +

# only show found search text
# find . -name "*.ts" -exec grep -h "debugFactory('loopback" {} +
```

Call

```sh
find_files_with_text.sh "*.ts" "'loopback:"
```

If you want to see :  filename searchText in the results, remove the `-h` in the script.
Right now it outputs ONLY the found search text :   searchText

Based on my search string above, I found:
```sh
 Files with extension : '*.ts'  
 that contains text   : ''loopback:' : 
----------------------------------------------

const debug = require('debug')('loopback:benchmark');
const debug = debugFactory('loopback:cron');
const debug = debugFactory('loopback:multi-tenancy:strategy:query');
const debug = debugFactory('loopback:multi-tenancy:strategy:header');
const debug = debugFactory('loopback:multi-tenancy:strategy:host');
const debug = debugFactory('loopback:multi-tenancy:strategy:jwt');
const debug = debugFactory('loopback:multi-tenancy:action');
const debug = require('debug')('loopback:example:acl');
const debug = debugFactory('loopback:boot:lb3app');
const debug = debugFactory('loopback:repository:belongs-to-accessor');
const debug = debugFactory('loopback:repository:belongs-to-helpers');
const debug = debugFactory('loopback:repository:has-one-helpers');
const debug = debugFactory('loopback:repository:has-one-repository-factory');
const debug = debugFactory('loopback:repository:has-many-repository-factory');
const debug = debugFactory('loopback:repository:has-many-inclusion-resolver');
const debug = debugFactory('loopback:repository:has-many-helpers');
const debug = debugFactory('loopback:repository:relation-helpers');
export const RELATIONS_KEY = 'loopback:relations';
const debug = debugFactory('loopback:repository:mixin');
>('loopback:model');
>('loopback:model-properties');
>('loopback:model-and-properties');
const debug = debugFactory('loopback:context:binding');
const debug = debugFactory('loopback:context:invocation');
const debug = debugFactory('loopback:context:interceptor');
const debug = debugModule('loopback:context:resolver');
const debug = debugFactory('loopback:context:binding-inspector');
   * const debug = debugFactory('loopback:context:application');
   *   return 'loopback:context:application';
    if (this.constructor === Context) return 'loopback:context';
const debug = debugFactory('loopback:context:subscription');
      expect(myCtx.debugFn.namespace).to.eql('loopback:context:mycontext');
const debug = debugModule('loopback:context:test');
const debug = debugFactory('loopback:context:interceptor-chain');
const debugSession = debugModule('loopback:context:resolver:session');
const debug = debugFactory('loopback:context:view');
const debug = debugFactory('loopback:core:lifecycle');
const debug = debugFactory('loopback:core:application');
const debugShutdown = debugFactory('loopback:core:application:shutdown');
const debugWarning = debugFactory('loopback:core:application:warning');
const debug = debugFactory('loopback:boot:booter-utils');
const debug = debugFactory('loopback:boot:base-artifact-booter');
const debug = debugFactory('loopback:boot:lifecycle-observer-booter');
const debug = debugFactory('loopback:boot:interceptor-booter');
const debug = debugFactory('loopback:boot:service-booter');
const debug = debugModule('loopback:boot:booter:application-metadata');
const debug = debugFactory('loopback:boot:model-api');
const debug = debugModule('loopback:boot:bootstrapper');
const debug = debugFactory('loopback:http-caching-proxy');
const debug = debugFactory('loopback:tsdocs');
const debug = debugFactory('loopback:tsdocs');
>('loopback:json-schema');
const debug = debugFactory('loopback:repository-json-schema:build-schema');
  'loopback:openapi3:metadata:controller-spec:deprecated',
const debug = require('debug')('loopback:openapi3:metadata:requestbody');
const debug = debugModule('loopback:openapi:spec-enhancer');
const debug = debugModule('loopback:openapi:spec-enhancer');
const debug = debugModule('loopback:openapi:spec-enhancer');
const debug = require('debug')('loopback:openapi3:metadata:controller-spec');
const debug = debugFactory('loopback:repository-tests');
const debug = debugFactory('loopback:authorization:interceptor');
const debug = debugModule('loopback:metadata:decorator');
const debug = debugFactory('loopback:boot:crud-rest');
const debug = debugFactory('loopback:openapi:spec-enhancer:consolidate');
const debug = debugFactory('loopback:openapi:spec-enhancer:info');
const debug = debugModule('loopback:rest:body-parser');
const debug = debugModule('loopback:rest:body-parser');
const debug = debugFactory('loopback:rest:parser');
const debug = debugFactory('loopback:openapi:spec-enhancer');
const debug = require('debug')('loopback:rest:sequence');
const debug = debugFactory('loopback:rest:server');
const debug = debugModule('loopback:rest:coercion');
const debug = debugModule('loopback:rest:coercion');
const debug = debugModule('loopback:rest:ajv');
const debug = debugModule('loopback:rest:validation');
const debug = require('debug')('loopback:rest:router:trie');
const debug = debugFactory('loopback:rest:routing-table');
const debug = require('debug')('loopback:rest:router:regexp');
const debug = debugFactory('loopback:rest:controller-route');

```

but feel free to adjust the search string

Reply:
Thanks!! I also found that some packages have different ways to define the string e.g [`@loopback:cli`](https://github.com/strongloop/loopback-next/blob/master/packages/cli/lib/debug.js) 😂 which I have no idea how to write a script for